KEP-4818: Relaxed validation for allowing zero in PreStop hook sleep action

sreeram-venkitesh opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
This PR adds a feature gate, `PodLifecycleSleepActionAllowZero`, to allow zero value for the sleep action in the PreStop container lifecycle hook.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes/enhancements/issues/4818

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Allows PreStop lifecycle handler's sleep action to have a zero value
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4818
```

kannon92 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3167,9 +3167,16 @@ func validateSleepAction(sleep *core.SleepAction, gracePeriod *int64, fldPath *f
-	if gracePeriod != nil && sleep.Seconds <= 0 || sleep.Seconds > *gracePeriod {
-		invalidStr := fmt.Sprintf("must be greater than 0 and less than terminationGracePeriodSeconds (%d)", *gracePeriod)
-		allErrors = append(allErrors, field.Invalid(fldPath, sleep.Seconds, invalidStr))
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodLifecycleSleepActionAllowZero) {
```
We aim to not have feature gates in the validation logic.

You could use the feature gate in the PodValidationOpts.

sreeram-venkitesh commented:
/sig node

sreeram-venkitesh commented:
@kannon92 Addressing your comment, I've updated the logic to use the feature gate set in `PodValidationOptions`. I've tried to describe the change in my second commit with the flow given below. The validation takes place in `validateSleepAction`, for which I've passed the PodValidationOptions down from `validatePodSpec`, through the different functions which ultimately end up calling `validateHandler` and thus `validateSleepAction`.

```
validatePodSpec
|__validateContainers
|__validateInitContainers
   |__validateLifecycle
   |    |__validateHandler
   |    |__validateSleepAction
   |__validateLivenessProbe
   |	  |__validateProbe
   |	     |__validateHandler
   |	        |__validateSleepAction
   |__validateReadinessProbe
   |	  |__validateProbe
   |	     |__validateHandler
   |	        |__validateSleepAction
   |__validateStartupProbe
   	  |__validateProbe
   	     |__validateHandler
   	        |__validateSleepAction
```

sreeram-venkitesh commented:
I'm working on fixing the tests.

kannon92 commented:
/hold

Please open a KEP first. You can leave this up but don't expect much reviews until the KEP is approved.

thockin commented on pkg/api/pod/util.go:
```diff
@@ -386,6 +386,7 @@ func GetValidationOptionsFromPodSpecAndMeta(podSpec, oldPodSpec *api.PodSpec, po
 		AllowNamespacedSysctlsForHostNetAndHostIPC:        false,
 		AllowNonLocalProjectedTokenPath:                   false,
 		AllowImageVolumeSource:                            utilfeature.DefaultFeatureGate.Enabled(features.ImageVolume),
+		AllowPodLifecycleSleepActionZeroValue:             utilfeature.DefaultFeatureGate.Enabled(features.PodLifecycleSleepActionAllowZero),
```
The usual pattern is to initialize this, then (below) do something like:

```
opts.AllowPodLifecycleSleepActionZeroValue = opts.AllowPodLifecycleSleepActionZeroValue || lifecycleSleepZeroUsed(podSpec)
```

Where:

```
func lifecycleSleepZeroUsed(spec *api.PodSpec) bool {
    for i := range spec.containers {
        c := &spec.Containers[i]
        if c.Lifecycle != nil {
            if c.Lifecycle.PostStart != nil && c.Lifecycle.PostStart.Sleep != nil && c.Lifecycle.PostStart.Sleep.Seconds == 0 {
                return true
            }
            if c.Lifecycle.PreStop != nil && c.Lifecycle.PreStop.Sleep != nil && c.Lifecycle.PreStop.Sleep.Seconds == 0 {
                return true
            }
        }
    }
    return false
}
```

There's a util function somewhere, I can't recall the name, that lets you easily repeat this for spec.Containers, .EphemeralContainers, and .InitContainers.  You should use that.

This allows us to respect existing pods which have set this value.  See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md for more

sreeram-venkitesh replied:
I believe the util function is [visitContainers](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/pod/util.go#L63).
https://github.com/kubernetes/kubernetes/blob/f153edf356324564396ae3f20538859783aadc76/pkg/api/pod/util.go#L63

It makes sense how by setting the following line `opts.AllowPodLifecycleSleepActionZeroValue = opts.AllowPodLifecycleSleepActionZeroValue || lifecycleSleepZeroUsed(podSpec)`, we can respect existing Pods with the value set to zero and make updates work and tolerate the zero value like you commented in the KEP Google doc. 

I can see a function with a similar pattern for the parent KEP too. This is really cool!
https://github.com/kubernetes/kubernetes/blob/f153edf356324564396ae3f20538859783aadc76/pkg/api/pod/util.go#L776-L781

I made the changes to add the `lifecycleSleepZeroUsed` function. I'm trying to test it locally but kind is throwing some errors. Let me take a look. I will push the code changes here as well as the KEP PR with the changes soon.

thockin commented on pkg/features/kube_features.go:
```diff
@@ -598,6 +598,13 @@ const (
+	// alpha: v1.32
+	//
+	// Allows zero value for sleep duration in SleepAction in container lifecycle hooks
+	PodLifecycleSleepActionAllowZero featuregate.Feature = "PodLifecycleSleepActionAllowZero"
```
Also need it in pkg/features/versioned_kube_features.go

thockin reviewed (commented):
This is nice, but it needs a unit test for validation and an e2e test for 0

sreeram-venkitesh commented:
@thockin PTAL! I've made it so that Pods with zero sleep duration can be updated and still continue to work and after the feature gate has been disabled. Also added some basic tests. 

CC: @SergeyKanzhelev @AxeZhan

SergeyKanzhelev commented on test/e2e/common/node/lifecycle_hook.go:
```diff
@@ -629,3 +629,36 @@ var _ = SIGDescribe(feature.PodLifecycleSleepAction, func() {
 	})
 })
+
+var _ = SIGDescribe(feature.PodLifecycleSleepActionAllowZero, func() {
```
were there any tests to disable feature gate in scope of a KEP?

sreeram-venkitesh replied:
> a test with feature.PodLifecycleSleepActionAllowZero disabled, for example, when fg is disabled, creating a pod with 0 duration will result in a error?

Ah yes, I have added a test case in `validation_test.go` for this:
https://github.com/kubernetes/kubernetes/blob/a0d4262276ee5db42e9104f7640d6d800c9e73d3/pkg/apis/core/validation/validation_test.go#L24180-L24188

SergeyKanzhelev replied:
ok, the feature is simple enough to not have an e2e_node test for it.

pacoxu commented on pkg/api/pod/util.go:
```diff
@@ -795,6 +798,28 @@ func podLifecycleSleepActionInUse(podSpec *api.PodSpec) bool {
+		return false
+	}
+	var inUse bool
+	VisitContainers(podSpec, AllContainers, func(c *api.Container, containerType ContainerType) bool {
```
Lifecycle is not allowed for ephemeral containers.
- However, `c.Lifecycle == nil` will return true for ephemeral containers.

sreeram-venkitesh replied:
Yes, for ephemeral containers `podLifecycleSleepActionZeroValueInUse` will return false after exiting from `VisitContainers`.

pacoxu commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3169,14 +3169,21 @@ func handlerFromLifecycle(lh *core.LifecycleHandler) commonHandler {
+		}
+	} else {
+		if gracePeriod != nil && sleep.Seconds <= 0 || sleep.Seconds > *gracePeriod {
+			invalidStr := fmt.Sprintf("must be greater than 0 and less than terminationGracePeriodSeconds (%d)", *gracePeriod)
```
Should we include a suggestion to include that you can enable the new FG to allow zero in the error message?

pacoxu commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3169,14 +3169,21 @@ func handlerFromLifecycle(lh *core.LifecycleHandler) commonHandler {
-		invalidStr := fmt.Sprintf("must be greater than 0 and less than terminationGracePeriodSeconds (%d)", *gracePeriod)
-		allErrors = append(allErrors, field.Invalid(fldPath, sleep.Seconds, invalidStr))
+	if opts.AllowPodLifecycleSleepActionZeroValue {
+		if gracePeriod != nil && sleep.Seconds < 0 || sleep.Seconds > *gracePeriod {
```
If gracePeriod can be nil, this should be
```
		if gracePeriod != nil && (sleep.Seconds < 0 || sleep.Seconds > *gracePeriod) {
```

pacoxu commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -24140,40 +24140,83 @@ func TestValidateSleepAction(t *testing.T) {
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			errs := validateSleepAction(tc.action, &tc.gracePeriod, fldPath)
+			errs := validateSleepAction(tc.action, &tc.gracePeriod, fldPath, PodValidationOptions{AllowPodLifecycleSleepActionZeroValue: tc.zeroValueEnabled})
```
All gracePeriod here are not nil in the UT.

pacoxu commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3169,14 +3169,21 @@ func handlerFromLifecycle(lh *core.LifecycleHandler) commonHandler {
+			allErrors = append(allErrors, field.Invalid(fldPath, sleep.Seconds, invalidStr))
+		}
+	} else {
+		if gracePeriod != nil && sleep.Seconds <= 0 || sleep.Seconds > *gracePeriod {
```
same as above.

sreeram-venkitesh commented:
@pacoxu I've added the changes from your review. For https://github.com/kubernetes/kubernetes/pull/127094#discussion_r1780552555 I believe it is fine for ephemeral containers since `podLifecycleSleepActionZeroValueInUse` returns false. This is based on the `podLifecycleSleepActionInUse` method which has the same behaviour. 

PTAL! Thank you!

pacoxu commented:
> /hold
> 
> Please open a KEP first. You can leave this up but don't expect much reviews until the KEP is approved.

/unhold 
as the KEP is merged

/priority important-soon
/triage accepted

SergeyKanzhelev commented on pkg/features/kube_features.go:
```diff
@@ -572,6 +572,13 @@ const (
+	// alpha: v1.32
+	//
+	// Allows zero value for sleep duration in SleepAction in container lifecycle hooks
+	PodLifecycleSleepActionAllowZero featuregate.Feature = "PodLifecycleSleepActionAllowZero"
```
have we discussed skipping alpha on this KEP? Just a side note

sreeram-venkitesh replied:
Tim had mentioned doing one alpha release and then moving to beta/GA in [this comment](https://github.com/kubernetes/enhancements/issues/3960#issuecomment-2209320166) in the parent KEP.

thockin replied:
Need to be able to roll-back, which would make saved data invalid.  Alpha->GA for this seems fine.

SergeyKanzhelev reviewed (approved):
/lgtm

e2e test is the best we can do I think

pacoxu commented on pkg/features/kube_features.go:
```diff
@@ -572,6 +572,13 @@ const (
 
+	// owner: @sreeram-venkitesh
+	// kep: http://kep.k8s.io/4818
+	// alpha: v1.32
```
We may remove this line as https://github.com/kubernetes/kubernetes/pull/128076.

pacoxu commented:
> This is nice, but it needs a unit test for validation and an e2e test for 0

Ping @thockin to take another look for api review as your comments should be addressed.

sreeram-venkitesh commented:
@pacoxu I've updated the kube_features.go file as per https://github.com/kubernetes/kubernetes/pull/128076 and the lgtm has been removed hence.

pacoxu commented:
/lgtm

SergeyKanzhelev commented on test/e2e/feature/feature.go:
```diff
@@ -267,6 +267,10 @@ var (
 	PodLifecycleSleepAction = framework.WithFeature(framework.ValidFeatures.Add("PodLifecycleSleepAction"))
 
+	// Owner: sig-node
+	// Marks a single test that tests Pod Lifecycle Sleep action with zero duration.
```
```suggestion
	// Marks a single test that tests Pod Lifecycle Sleep action with zero duration. Requires feature gate PodLifecycleSleepActionAllowZero to be enabled.
```

SergeyKanzhelev reviewed (approved):
nothing much from the node side. And already approved. API side changes look reasonable.

Please squash commits before lgtm

sreeram-venkitesh committed:
Clean kube_features.go

Added tests, info about new feature gate in error message, fixes from review

Added basic e2e test

Added unit tests

Ran hack/update-featuregates.sh

Tolerate updates to existing resources after disabling feature gate

Added feature gate to versioned_kube_features.go

Fixed existing tests

Use PodValidationOptions for validation instead of using feature gate directly

Relaxed validation for allowing zero in prestop hook sleep action

SergeyKanzhelev reviewed (approved):
/lgtm

SergeyKanzhelev commented:
/assign @thockin

sreeram-venkitesh commented:
/retest

thockin commented:
/approve

sreeram-venkitesh commented:
Thanks Tim!

@SergeyKanzhelev lgtm was removed after rebase

thockin commented:
/lgtm
